### PR TITLE
feat: Add ability to override file extension resolution

### PR DESCRIPTION
**Pull request by Shepless**
FuseBox can now be configured to look for customised extensions ahead of standard ones. This is
useful for platform specific builds. For example:

```js
FuseBox.init({
   extensionOverrides: ['.desktop.js']
});
```

With the above configuration FuseBox will favour loading contents from files with a `.desktop.js` extension over `.js` extension. If `.desktop.js` does not exist then normal resolution will occur. This a essentially a port of the Webpack functionality for [resolve extensions](https://webpack.js.org/configuration/resolve/#resolve-extensions).

In Summary:
- Added Functionality
- Added Unit Tests
- Added Documentation

### CODE REVIEW

File: src/core/ExtensionOverrides.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  public add(override: string) {
+    if (this.isValid(override)) {
+      this.overrides.push(override);
+    }
+  }
+
+  public setOverrideFileInfo(file: File): string {
+    if (this.overrides.length === 0) {
+      return null;
```
Comment (nchanged) [inline src/core/ExtensionOverrides.ts]: nitpick: better to return `undefined` (nothing)

File: src/core/File.ts [hunk 2]
```diff
@@ -159,10 +161,14 @@ export class File {
...
             this.params = info.params;
         }
+
         this.absPath = info.absPath;
+
         if (this.absPath) {
             this.relativePath = ensureFuseBoxPath(path.relative(this.context.appRoot, this.absPath));
         }
+
+        this.context.extensionOverrides.setOverrideFileInfo(this);
```
Comment (nchanged) [inline src/core/File.ts]: What is it doing in alternative content? Shouldn't it be at `consume` level? That one limits many plugins

File: src/core/ExtensionOverrides.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    const fileInfo = path.parse(file.info.absPath);
+
+    for (let overrideExtension of this.overrides) {
+      const overridePath = path.resolve(fileInfo.dir, `${fileInfo.name}${overrideExtension}`);
+
+      if (overrideExtension.indexOf(fileInfo.ext) > -1 && fs.existsSync(overridePath)) {
+        file.absPath = file.info.absPath = overridePath;
```
Comment (nchanged) [inline src/core/ExtensionOverrides.ts]: i don't think we should override absPath ( my bad,  I told you that )
we should map it in the context instead for the watcher

File: src/tests/BundleApi.test.ts [hunk 4]
```diff
@@ -135,6 +135,34 @@ export class BundleApiTest {
...
+    "Should not share extension overrides across bundles"() {
+        const fuse = createFuse();
+        const bundle1 = fuse.bundle("app")
+            .extensionOverrides(".foo.js", ".foo.json");
+        const bundle2 = fuse.bundle("app")
+            .extensionOverrides(".bar.js", ".bar.json");
+
+        should(bundle1.context.extensionOverrides.overrides).deepEqual([".foo.js", ".foo.json"]);
+        should(bundle2.context.extensionOverrides.overrides).deepEqual([".bar.js", ".bar.json"]);
+    }
```
Comment (nchanged) [inline src/tests/BundleApi.test.ts]: Not enough tests, we should make similar tests additionally

https://github.com/fuse-box/fuse-box/blob/8586d90baac5326f849aadf8aca82377a7e33764/src/tests/plugins/CSSPlugin.test.ts#L131-L147

To cover all the cases, not just VuePlugin

File: src/core/ExtensionOverrides.ts (hunk 3 above)
Comment (nchanged) [reply src/core/ExtensionOverrides.ts]: Using HMR_FILE_REQUIRED
https://github.com/fuse-box/fuse-box/blob/8586d90baac5326f849aadf8aca82377a7e33764/src/plugins/stylesheet/CSSplugin.ts#L132-L137

**Comment by TombolaShepless** [comment]
@nchanged cheers for the feedback - will go through it tonight and update!

File: src/core/ExtensionOverrides.ts (hunk 1 above)
Comment (nchanged) [reply src/core/ExtensionOverrides.ts]: Also let's add a condition if it belongs to the actual project (vendors ignored)
`file.belongsToProject()`

File: src/core/File.ts (hunk 2 above)
Comment (Shepless) [reply src/core/File.ts]: It doesn't just do this in alternative content. We do this work in the `File` constructor so it happens at the lowest level possible. All testing I have done doesn't look to have impacted.

File: src/tests/BundleApi.test.ts (hunk 4 above)
Comment (Shepless) [reply src/tests/BundleApi.test.ts]: Im not sure I understand? The tests are in the bundle? I should probably add some for `File` though for sure.

File: src/core/File.ts (hunk 2 above)
Comment (nchanged) [reply src/core/File.ts]: Then why is it happening on `addAlternativeContent` ?

File: src/tests/BundleApi.test.ts (hunk 4 above)
Comment (nchanged) [reply src/tests/BundleApi.test.ts]: No, I just gave you an example. You can create test bundle in a test, with the actual file structure and stuff.

File: src/core/File.ts (hunk 2 above)
Comment (Shepless) [reply src/core/File.ts]: Its not?
screen shot 2017-10-16 at 20 43 47

Comment (nchanged) [reply src/core/File.ts]: sorry man, i am tired ;-))

Comment (Shepless) [reply src/core/File.ts]: hehe no worries bud.

**Comment by Shepless** [comment]
Closing as the rebase went horribly wrong haha! I've sorted on a new branch.
